File: 1856-maximum-subarray-min-product/1856-maximum-subarray-min-product.java
class Solution {
    static final int MOD =  1000000007;
    void nse(int[] suf , int[] nums){
        int n = nums.length;
        Stack<Integer> st = new Stack<>();
        st.push(n - 1);
        suf[n - 1] = n;
        for(int i = n - 2 ; i >= 0 ;i--){
            while(!st.isEmpty() && nums[st.peek()] >= nums[i]){
                st.pop();
            }
            if(st.isEmpty()) suf[i] = n;
            else{
                suf[i] = st.peek();
            }
            st.push(i); 
        }
    }
    void pse(int[] pref , int[] nums){
        int n = nums.length;
        Stack<Integer> st = new Stack<>();
        st.push(0);
        pref[0] = -1;
        for(int i = 1 ; i < n ; i++){
            while(!st.isEmpty() && nums[st.peek()] >= nums[i]){
                st.pop();
            }
            if(st.isEmpty()) pref[i] = -1;
            else{
                pref[i] = st.peek();
            } 
            st.push(i);
        }
    }
    public int maxSumMinProduct(int[] nums) {
        int n = nums.length;
        int[] suf = new int[n];
        int[] pref = new int[n];
        nse(suf , nums);
        pse(pref , nums);

        long[] ps = new long[n + 1];
        for(int i = 0  ; i < n ; i++){
            ps[i + 1] = nums[i] + ps[i];
        }
        
        long maxi  = 0;
        for(int i = 0 ; i < n ; i++){
            int s = suf[i];
            int p = pref[i];
            long sum = ps[s] - ps[p + 1];
            long prod = 1L * nums[i] * sum;
            maxi = Math.max(maxi , prod);
        }
        return (int)(maxi % MOD);
    }
}

File: 1856-maximum-subarray-min-product/1856-maximum-subarray-min-product.cpp
class Solution {
public:
    #define MOD 1000000007
    void pse(vector<int>& prefMin , vector<int>& nums){
        int n = nums.size();
        prefMin[0] = -1;
        stack<int> st;
        st.push(0);
        for(int i = 1 ; i < n ; i++){
            while(!st.empty() && nums[st.top()] >= nums[i]){
                st.pop();
            }
            if(st.empty()) prefMin[i] = -1;
            else{
                prefMin[i]  = st.top();
            }
            st.push(i);
        }
    }
    void nse(vector<int>& suffMin , vector<int>& nums){
        int n = nums.size();
        suffMin[n - 1] = n;
        stack<int> st;
        st.push( n - 1);
        for(int i = n - 2 ; i >= 0 ; i--){
            while(!st.empty() && nums[st.top()] >= nums[i]){
                st.pop();
            }
            if(st.empty()) suffMin[i] = n;
            else{
                suffMin[i]  = st.top();
            }
            st.push(i);
        }
    }
    int maxSumMinProduct(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefMin(n) , suffMin(n);
        pse(prefMin , nums);
        nse(suffMin , nums);

        long long maxi = 0;

        vector<long long> prefs(n + 1 , 0);
        for(int i = 0 ; i < n ; i++){
            prefs[i + 1] = prefs[i] + nums[i]; 
        }

        for(int i = 0 ; i < n ; i++){
            int p = prefMin[i] ;
            int s = suffMin[i];
            long long sum = (prefs[s] - prefs[p + 1]);
            long long prod = 1LL * nums[i] * sum;
            maxi = max(maxi , prod);
        }
        return maxi % MOD;
    }
};